docs/design/l2-peer-protocol-spec.md

```diff
@@ -0,0 +1,507 @@
+    , htlcAmountRange :: Range Amount
+    , totalHtlcAmountRange :: Range Amount
+    , minGiftAmount :: Amount                -- Minimum "service fee" required
+    , minGiftPercent :: Integer              -- Used above the `minGiftAmount`
+    , peerFundingRange :: Range Amount       -- Minimum funding required from peer
+    , respondPeriodRange :: Range Milliseconds
+    , termsAndConditions :: Maybe Text
+    }
+
+data QueryError
+  = QueryError
+      { supportedCurrencies :: [(Currency, Text)]
+      }
+```
+
+This mechanism have obvious limitations:
+
+- Provided ranges can have internal dependencies - for example `responsePeriod`
+  can depend on the funds locked in a channel so even though the `opener` can
+  provide the final channel proposition which fullfiles per field requirements
+  it can still be rejected by the counterparty. More involved policies can use
+  `termsAndConditions` field to explain the details for the user.
+- The list is not exhaustive - it can be extended in the future.
+
+#### Opening and Staging
+
+> QUESTION:
+>
+> - Shoudl we expand the failure scenarios - transaction can timeout etc.?
+> - Should we discuss the transaction validity interval and in the future fee
+>   bumping etc.?
+
+The opening consists of a simple negotiation where the `openner` proposes the
+channel parameters and the `non-opener` accepts them with possible adjustments.
+Then the `opener` finally confirms the channel by sending a signed transaction
+which can be submitted by both parties. After this step both parties should
+confirm about the channel creation and start the operation or report an error
+and drop the channel.
+
+```haskell
+data Open
+  = Open
+    { currency :: Currency
+    , fundingAmount :: Amount
+    , giftAmount :: Amount
+    , respondPeriod :: Milliseconds
+    , minimumDepth :: Integer
+    , minHtlcAmount :: Amount
+    , maxHtlcAmount :: Range Amount
+    , maxTotalHtlcAmount :: Range Amount
+    , maxHtlcCount :: Integer
+    , pubKey :: Ed25519PubKey
+    }
+
+data Accept
+  = Accept
+    { respondPeriod :: Milliseconds
+    , minimumDepth :: Integer
+    , minHtlcAmount :: Amount
+    , maxHtlcAmount :: Amount
+    , maxTotalHtlcAmount :: Amount
+    , maxHtlcCount :: Integer
+    , pubKey :: Ed25519PubKey
+    }
+
+type Cbor = ByteString
+
+type Transaction = Cbor
+
+data ChannelStaging
+  = ChannelStaging
+    { tx :: Transaction
+    }
+
+data ChannelStaged = ChannelStaged
+```
+
+### Operating
+
+During this phase all the messages in the current version of the protocol are
+exchanged in a fully asynchronous and pipelined manner. We tag all the
+conversation with unique identifier which is repeated across the conversation.
+Both parties keep an indpendent sequence number which is used for an `id`. To
+avoid collisions `opener` uses even numbers and `non-opener` uses odd numbers.
+
+#### Chenque Management
+
+This flow consists of a single request and acknowledge pairs:
+
+- `AddCheque`/`ChequeAdded`
```

Comment:
How do you feel about removing `Cheque` from everyone of these? 

The context will be abundantly clear that we are reffering to cheques.